op-node/rollup/sync/start.go

```diff
@@ -40,164 +27,192 @@ import (
 			// If the reorg depth is too large, something is fishy.
 			// This can legitimately happen if L1 goes down for a while. But in that case,
 			// restarting the L2 node with a bigger configured MaxReorgDepth is an acceptable
 			// stopgap solution.
-			// Currently this can also happen if the L2 node is down for a while, but in the future
-			// state sync should prevent this issue.
-			return eth.L2BlockRef{}, eth.L2BlockRef{}, TooDeepReorgErr
+			return nil, fmt.Errorf("%w: traversed back to L2 block %s, but too deep compared to previous unsafe block %s", TooDeepReorgErr, n, prevUnsafe)
 		}
-	}
-
-	// Loop 2. Walk from the L1 origin of the `n` block (*) back to the L1 block that starts the
-	// sequencing window ending at that block. Instead of iterating on L1 blocks, we actually
-	// iterate on L2 blocks, because we want to find the safe L2 head, i.e. the highest L2 block
-	// whose L1 origin is the start of the sequencing window.
 
-	// (*) `n` being at this stage the highest L2 block whose L1 origin is canonical.
-
-	// Depth counter: we need to walk back `seqWindowSize` L1 blocks in order to find the start
-	// of the sequencing window.
-	depth := uint64(1)
+		// If we don't have a usable unsafe head, then set it
+		if result.Unsafe == (eth.L2BlockRef{}) {
+			result.Unsafe = n
+		}
 
-	// Before entering the loop: `prevL1OriginHash == n.L1Origin.Hash`
-	// The original definitions of `n` and `prevL1OriginHash` still hold.
-	for {
-		// Advance depth if we change to a different (earlier) epoch.
-		if n.L1Origin.Hash != prevL1OriginHash {
-			depth++
-			prevL1OriginHash = n.L1Origin.Hash
+		if ahead {
+			// keep the unsafe head if we can't tell if its L1 origin is canonical or not yet.
+		} else if l1Block.Hash == n.L1Origin.Hash {
+			// if L2 matches canonical chain, even if unsafe,
+			// then we can start finding a span of L1 blocks to cover the sequence window,
+			// which may help avoid rewinding the existing safe head unnecessarily.
+			if highestL2WithCanonicalL1Origin == (eth.L2BlockRef{}) {
+				highestL2WithCanonicalL1Origin = n
+			}
+		} else {
+			// L1 origin nor ahead of L1 head nor canonical, discard previous candidate and keep looking.
+			result.Unsafe = eth.L2BlockRef{}
+			highestL2WithCanonicalL1Origin = eth.L2BlockRef{}
 		}
 
-		// Found an L2 block whose L1 origin is the start of the sequencing window.
-		// Note: We also ensure that we are on the block number with the 0 seq number.
-		// This is a little hacky, but kinda works. The issue is about where the
-		// batch queue should start building.
-		if depth == seqWindowSize && n.SequenceNumber == 0 {
-			return highestPlausibleCanonicalOrigin, n, nil
+		// If the L2 block is at least as old as the previous safe head, and we have seen at least a full sequence window worth of L1 blocks to confirm
+		// TODO discuss with Josh
+		if n.Number <= result.Safe.Number && n.Number+cfg.SeqWindowSize < highestL2WithCanonicalL1Origin.Number && n.SequenceNumber == 0 {
```

Comment:
seqNumber check almost seems like a return check over finding the safehead check.
I also need to think more if we actually need the full sequence window check here. Maybe, but it's a weird to me way of thinking about it.